File: Parallel Heat/heat.jl
using Base.Threads
using Plots
using LinearAlgebra
using Printf

function solve_heat_parallel()
    # Parameters
    α = 0.1     # Thermal diffusivity
    L = 100000.0     # Domain length
    n = 1000000     # Spatial points
    h = L / (n - 1)
    T = 0.1     # Total time
    dt = T / 1000
    nt = Int(T / dt)
    
    # Initialize arrays
    x = range(0, L, length=n)
    u_old = sin.(π .* x)
    u_new = similar(u_old)
    
    # Precompute coefficients
    factor = α * dt / h^2
    
    # Time stepping loop with multithreading
    @time for _ in 1:nt
        @threads for i in 2:n-1
            u_new[i] = u_old[i] + factor * (u_old[i+1] - 2u_old[i] + u_old[i-1])
        end
        # Boundary conditions
        u_new[1] = 0.0
        u_new[end] = 0.0
        u_old, u_new = u_new, u_old  # Swap arrays
    end
    
    # Analytical solution
    analytical = exp(-α * π^2 * T) .* sin.(π .* x)
    
    # Calculate L2 error
    l2_error = norm(u_old - analytical) * sqrt(h)
    @printf "L2 Error: %.4e\n" l2_error
    
    # Plot results
    plot(x, u_old, label="Numerical", linewidth=2, xlabel="x", ylabel="u(t, x)", 
         title="1D Heat Equation Solution at t=0.1", legend=:topright)
    plot!(x, analytical, label="Analytical", linestyle=:dash, linewidth=2)
    savefig("heat_equation_julia.png")
end

# Run the solver
solve_heat_parallel()

File: Parallel Heat/heat.py
import numpy as np
import matplotlib.pyplot as plt
from numba import njit, prange
import time

# Parameters
alpha = 0.1   # Thermal diffusivity
L = 100000.0      # Length of the domain
n = 10000000       # Number of spatial points
h = L / (n - 1)  # Spatial step size
T = 0.1       # Total simulation time
dt = T / 1000 # Time step size
nt = 1000  # Number of time steps

# Spatial grid
x = np.linspace(0, L, n)
# Initial condition
u0 = np.sin(np.pi * x)

@njit(parallel=True)
def solve_heat(u, nt, dt, h, alpha):
    """
    Solve the 1D heat equation using finite differences and Euler method.
    Parallel implementation using Numba.
    """
    n = u.shape[0]
    u_new = np.empty_like(u)
    for _ in range(nt):
        for i in prange(n):
            if i == 0 or i == n-1:
                u_new[i] = 0.0  # Dirichlet boundary conditions
            else:
                # Finite difference scheme
                laplacian = (u[i+1] - 2*u[i] + u[i-1]) / h**2
                u_new[i] = u[i] + alpha * dt * laplacian
        u[:] = u_new[:]  # Update for next iteration
    return u

# Solve the heat equation
u_numerical = u0.copy()
start_time = time.perf_counter()
u_numerical = solve_heat(u_numerical, nt, dt, h, alpha)
end_time = time.perf_counter()
elapsed_time = end_time - start_time
print(elapsed_time)

# Analytical solution
t_final = T
analytical = np.exp(-alpha * np.pi**2 * t_final) * np.sin(np.pi * x)

# Plotting results
plt.figure(figsize=(10, 6))
plt.plot(x, u_numerical, label='Numerical Solution', linewidth=2)
plt.plot(x, analytical, '--', label='Analytical Solution', linewidth=2)
plt.xlabel('x', fontsize=12)
plt.ylabel('u(t, x)', fontsize=12)
plt.legend(fontsize=12)
plt.title('1D Heat Equation Solution at t=0.1', fontsize=14)
plt.grid(True)
plt.show()

# Calculate and print L2 error
error = np.sqrt(h * np.sum((u_numerical - analytical)**2))
#print(f"L2 Error: {error:.6e}")
